etna/transforms/math/lags.py

```diff
@@ -96,3 +100,205 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
+
+        self._freq: Optional[str] = None
+        self._created_regressors: Optional[List[str]] = None
+        self._exog_shifts: Optional[Dict[str, int]] = None
+        self._exog_last_date: Optional[Dict[str, pd.Timestamp]] = None
+        self._filter_out_columns = {"target"}
+
+        if isinstance(lag, int):
+            if lag <= 0:
+                raise ValueError(f"{self.__class__.__name__} works only with positive lags values, {lag} given")
+            self.lag = lag
+
+        else:
+            if horizon is None:
+                raise ValueError("Value of `horizon` should be specified when using `auto`!")
+
+            if horizon < 1:
+                raise ValueError(f"{self.__class__.__name__} works only with positive horizon values, {horizon} given")
+
+            self.horizon = horizon
+            self._auto = True
+
+    def _save_exog_last_date(self, df_exog: pd.DataFrame):
+        """Save last available date of each exogenous variable."""
+        exog_names = set(df_exog.columns.get_level_values("feature"))
+
+        self._exog_last_date = dict()
+        for name in exog_names:
+            feature = df_exog.loc[:, pd.IndexSlice[:, name]]
+
+            na_mask = pd.isna(feature).any(axis=1)
+            last_date = feature.index[~na_mask].max()
+
+            self._exog_last_date[name] = last_date
+
+    def fit(self, ts: TSDataset) -> "ExogShiftTransform":
+        """Fit the transform.
+
+        Parameters
+        ----------
+        ts:
+            Dataset to fit the transform on.
+
+        Returns
+        -------
+        :
+            The fitted transform instance.
+        """
+        self._freq = ts.freq
+        df_exog = ts.df_exog
+
+        if df_exog is not None:
+            self._save_exog_last_date(df_exog=df_exog)
+
+        else:
+            self._exog_last_date = {}
+
+        super().fit(ts=ts)
+
+        return self
+
+    def _fit(self, df: pd.DataFrame) -> "ExogShiftTransform":
+        """Estimate shifts for exogenous variables.
+
+        Parameters
+        ----------
+        df:
+            dataframe with data.
+
+        Returns
+        -------
+        :
+            Fitted `ExogShiftTransform` instance.
+        """
+        feature_names = self._get_feature_names(df=df)
+
+        self._exog_shifts = dict()
+        self._created_regressors = []
+
+        for feature_name in feature_names:
+            shift = self._estimate_shift(df=df, feature_name=feature_name)
+            self._exog_shifts[feature_name] = shift
+
+            if shift > 0:
+                self._created_regressors.append(f"{feature_name}_shift_{shift}")
+
+        return self
+
+    def _get_feature_names(self, df: pd.DataFrame) -> List[str]:
+        """Return the names of exogenous variables."""
+        if self._exog_last_date is not None and len(self._exog_last_date) > 0:
+            feature_names = list(self._exog_last_date.keys())
+
+        else:
+            feature_names = []
```

Comment:
Could we really reach this `else` clause?

Reply:
Yes, for example when dataset has no exog varibales.

Reply:
Can we rewrite it like this?
```python
        if self._exog_last_date is not None:
            feature_names = list(self._exog_last_date.keys())

        else:
            feature_names = []
```

Because we have a guarantee that if `self._exog_last_date` isn't `None` it is dict.

It also can't be called with `self._exog_last_date is None` unless it is called directly.